File: packages/auth/src/__tests__/supabase-auth-adapter.test.ts
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import {
  SupabaseAuthAdapter,
  type AuthStateChangeEvent,
  type SupabaseAuthClient,
  type SupabaseAuthSession,
  type UserProfileFetcher,
  type UserProfileRow,
} from "../supabase-auth-adapter.js";
import { GOOGLE_SCOPES } from "../types.js";

const REFRESH_URL = "https://example.supabase.co/functions/v1/auth/refresh-provider-token";

const PROFILE_ROW: UserProfileRow = {
  user_id: "550e8400-e29b-41d4-a716-446655440000",
  display_name: "Vgaka",
  storage_provider: "GOOGLE_DRIVE",
  calendar_provider: "GOOGLE_CALENDAR",
  ai_prefs: {
    chatProvider: "OLLAMA",
    chatModel: "qwen2.5:7b",
    tagProvider: "OLLAMA",
    tagModel: "llama3.2:3b",
    embeddingProvider: "OLLAMA",
    embeddingModel: "nomic-embed-text",
  },
  analytics_optout: false,
};

const SESSION: SupabaseAuthSession = {
  user: { id: PROFILE_ROW.user_id },
  access_token: "jwt-abc",
};

interface FakeClient extends SupabaseAuthClient {
  fire(event: AuthStateChangeEvent, session: SupabaseAuthSession | null): Promise<void>;
  readonly signInCalls: Array<Parameters<SupabaseAuthClient["signInWithOAuth"]>[0]>;
  signInError: { message: string; name?: string } | null;
  sessionFromGetSession: SupabaseAuthSession | null;
  unsubscribeCalls: number;
}

function makeFakeClient(): FakeClient {
  let cb: ((event: AuthStateChangeEvent, session: SupabaseAuthSession | null) => void) | null =
    null;
  const client: FakeClient = {
    signInCalls: [],
    signInError: null,
    sessionFromGetSession: null,
    unsubscribeCalls: 0,
    async signInWithOAuth(opts) {
      client.signInCalls.push(opts);
      return { data: null, error: client.signInError };
    },
    async signOut() {
      return { error: null };
    },
    async getSession() {
      return { data: { session: client.sessionFromGetSession }, error: null };
    },
    onAuthStateChange(handler) {
      cb = handler;
      return {
        data: {
          subscription: {
            unsubscribe: () => {
              client.unsubscribeCalls += 1;
            },
          },
        },
      };
    },
    async fire(event, session) {
      cb?.(event, session);
      // Let microtasks from the handler settle (it calls async profile fetcher).
      await Promise.resolve();
      await Promise.resolve();
    },
  };
  return client;
}

interface Harness {
  adapter: SupabaseAuthAdapter;
  client: FakeClient;
  profileFetcher: ReturnType<typeof vi.fn<UserProfileFetcher>>;
  fetcher: ReturnType<typeof vi.fn<typeof fetch>>;
}

function makeHarness(): Harness {
  const client = makeFakeClient();
  const profileFetcher = vi.fn(async () => PROFILE_ROW);
  const fetcher = vi.fn(
    async () =>
      new Response(JSON.stringify({ ok: true, data: { access_token: "ya29.fresh" } }), {
        status: 200,
        headers: { "Content-Type": "application/json" },
      }),
  );
  const adapter = new SupabaseAuthAdapter({
    client,
    userProfileFetcher: profileFetcher,
    refreshProviderTokenUrl: REFRESH_URL,
    redirectTo: "https://app.example.com/auth/callback",
    fetcher,
    signInTimeoutMs: 50,
  });
  return { adapter, client, profileFetcher, fetcher };
}

describe("SupabaseAuthAdapter.signInWithGoogle", () => {
  let harness: Harness;
  beforeEach(() => {
    harness = makeHarness();
  });
  afterEach(() => {
    harness.adapter.dispose();
  });

  it("requests the spec'd Google scopes via signInWithOAuth", async () => {
    const signIn = harness.adapter.signInWithGoogle();
    await harness.client.fire("SIGNED_IN", SESSION);
    const result = await signIn;
    expect(harness.client.signInCalls).toHaveLength(1);
    expect(harness.client.signInCalls[0]?.provider).toBe("google");
    expect(harness.client.signInCalls[0]?.options?.scopes).toBe(GOOGLE_SCOPES.join(" "));
    expect(harness.client.signInCalls[0]?.options?.redirectTo).toBe(
      "https://app.example.com/auth/callback",
    );
    expect(result.kind).toBe("Success");
  });

  it("resolves Success once the auth-state change carries a session", async () => {
    const signIn = harness.adapter.signInWithGoogle();
    await harness.client.fire("SIGNED_IN", SESSION);
    const result = await signIn;
    expect(result.kind).toBe("Success");
    if (result.kind === "Success") {
      expect(result.user.displayName).toBe("Vgaka");
    }
  });

  it("maps a user-cancelled provider response to AuthResult.Cancelled", async () => {
    harness.client.signInError = { message: "user_cancelled_login", name: "OAuthCancelled" };
    const result = await harness.adapter.signInWithGoogle();
    expect(result.kind).toBe("Cancelled");
  });

  it("maps any other provider error to AuthResult.Error", async () => {
    harness.client.signInError = { message: "redirect_uri_mismatch" };
    const result = await harness.adapter.signInWithGoogle();
    expect(result.kind).toBe("Error");
  });

  it("times out if no auth-state event arrives", async () => {
    const result = await harness.adapter.signInWithGoogle();
    expect(result.kind).toBe("Cancelled");
    if (result.kind === "Cancelled") {
      expect(result.reason).toMatch(/timed out/i);
    }
  });
});

describe("SupabaseAuthAdapter session lifecycle", () => {
  it("subscribers receive the initial null state and subsequent updates", async () => {
    const harness = makeHarness();
    const seen: Array<string | null> = [];
    harness.adapter.subscribeToUser((u) => seen.push(u?.displayName ?? null));
    await harness.client.fire("SIGNED_IN", SESSION);
    await harness.client.fire("SIGNED_OUT", null);
    expect(seen).toEqual([null, "Vgaka", null]);
    harness.adapter.dispose();
  });

  it("currentUser exposes the latest snapshot", async () => {
    const harness = makeHarness();
    expect(harness.adapter.currentUser()).toBeNull();
    await harness.client.fire("SIGNED_IN", SESSION);
    expect(harness.adapter.currentUser()?.displayName).toBe("Vgaka");
    harness.adapter.dispose();
  });

  it("signOut throws when Supabase reports a failure", async () => {
    const client = makeFakeClient();
    client.signOut = async () => ({ error: { message: "boom" } });
    const adapter = new SupabaseAuthAdapter({
      client,
      userProfileFetcher: async () => PROFILE_ROW,
      refreshProviderTokenUrl: REFRESH_URL,
    });
    await expect(adapter.signOut()).rejects.toThrow(/boom/);
    adapter.dispose();
  });

  it("dispose unsubscribes from the auth-state stream", () => {
    const harness = makeHarness();
    expect(harness.client.unsubscribeCalls).toBe(0);
    harness.adapter.dispose();
    expect(harness.client.unsubscribeCalls).toBe(1);
  });
});

describe("SupabaseAuthAdapter.providerAccessToken", () => {
  it("returns null when there is no Supabase session", async () => {
    const harness = makeHarness();
    const token = await harness.adapter.providerAccessToken("GOOGLE_DRIVE");
    expect(token).toBeNull();
    expect(harness.fetcher).not.toHaveBeenCalled();
    harness.adapter.dispose();
  });

  it("hits the Edge Function with the user JWT and unwraps the access_token", async () => {
    const harness = makeHarness();
    harness.client.sessionFromGetSession = SESSION;
    const token = await harness.adapter.providerAccessToken("GOOGLE_DRIVE");
    expect(token).toBe("ya29.fresh");
    expect(harness.fetcher).toHaveBeenCalledTimes(1);
    const call = harness.fetcher.mock.calls[0];
    if (!call) throw new Error("expected exactly one fetch call");
    const [url, init] = call;
    expect(url).toBe(REFRESH_URL);
    const reqInit = init as RequestInit;
    expect(reqInit.headers).toMatchObject({
      Authorization: "Bearer jwt-abc",
      "Content-Type": "application/json",
    });
    expect(JSON.parse(String(reqInit.body))).toEqual({ provider: "GOOGLE_DRIVE" });
    harness.adapter.dispose();
  });

  it("returns null when the Edge Function returns a non-2xx", async () => {
    const harness = makeHarness();
    harness.client.sessionFromGetSession = SESSION;
    harness.fetcher.mockResolvedValueOnce(new Response("nope", { status: 500 }));
    const token = await harness.adapter.providerAccessToken("GOOGLE_DRIVE");
    expect(token).toBeNull();
    harness.adapter.dispose();
  });

  it("returns null when the Edge Function returns an error shape", async () => {
    const harness = makeHarness();
    harness.client.sessionFromGetSession = SESSION;
    harness.fetcher.mockResolvedValueOnce(
      new Response(JSON.stringify({ ok: false, error: { code: "X", message: "y" } }), {
        status: 200,
      }),
    );
    const token = await harness.adapter.providerAccessToken("GOOGLE_DRIVE");
    expect(token).toBeNull();
    harness.adapter.dispose();
  });
});

describe("SupabaseAuthAdapter.accessToken", () => {
  it("returns null when no session is available", async () => {
    const harness = makeHarness();
    expect(await harness.adapter.accessToken()).toBeNull();
    harness.adapter.dispose();
  });

  it("returns the JWT when the session is available", async () => {
    const harness = makeHarness();
    harness.client.sessionFromGetSession = SESSION;
    expect(await harness.adapter.accessToken()).toBe("jwt-abc");
    harness.adapter.dispose();
  });
});

File: packages/android/auth-android/src/test/kotlin/app/notes/auth/SupabaseAuthAdapterTest.kt
package app.notes.auth

import app.notes.domain.CalendarProviderId
import app.notes.domain.ExternalProviderId
import app.notes.domain.StorageProviderId
import app.notes.domain.UserId
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class SupabaseAuthAdapterTest {
    private val userId = UserId("550e8400-e29b-41d4-a716-446655440000")
    private val session = StoredSession(
        userId = userId,
        accessToken = "jwt-abc",
        refreshToken = "rt-abc",
        expiresAtEpochSeconds = 9_999_999_999,
    )
    private val profile = UserProfileRow(
        userId = userId,
        displayName = "Vgaka",
        storageProvider = StorageProviderId.GoogleDrive,
        calendarProvider = CalendarProviderId.GoogleCalendar,
        aiPrefs = UserProfileRow.defaultPrefs(),
        analyticsOptOut = false,
    )

    @Test
    fun `signInWithGoogle stores session and returns Success`() = runTest {
        val http = FakeHttp(oauth = OAuthOutcome.Authenticated(session), profile = profile)
        val store = FakeStore()
        val adapter = SupabaseAuthAdapter(http, store)
        val result = adapter.signInWithGoogle()
        assertTrue(result is AuthResult.Success)
        assertEquals("Vgaka", (result as AuthResult.Success).user.displayName)
        assertEquals(session, store.current())
        assertEquals(listOf(GoogleScopes.REQUIRED), http.scopesRequested)
    }

    @Test
    fun `signInWithGoogle maps cancellation`() = runTest {
        val http = FakeHttp(oauth = OAuthOutcome.Cancelled("user closed tab"))
        val adapter = SupabaseAuthAdapter(http, FakeStore())
        val result = adapter.signInWithGoogle()
        assertTrue(result is AuthResult.Cancelled)
    }

    @Test
    fun `signInWithGoogle maps failure`() = runTest {
        val http = FakeHttp(oauth = OAuthOutcome.Failed("boom", null))
        val adapter = SupabaseAuthAdapter(http, FakeStore())
        val result = adapter.signInWithGoogle()
        assertTrue(result is AuthResult.Error)
    }

    @Test
    fun `signInWithGoogle returns Error when profile row missing`() = runTest {
        val http = FakeHttp(oauth = OAuthOutcome.Authenticated(session), profile = null)
        val adapter = SupabaseAuthAdapter(http, FakeStore())
        val result = adapter.signInWithGoogle()
        assertTrue(result is AuthResult.Error)
    }

    @Test
    fun `signOut clears the session and notifies subscribers`() = runTest {
        val http = FakeHttp(oauth = OAuthOutcome.Authenticated(session), profile = profile)
        val store = FakeStore(initial = session)
        val adapter = SupabaseAuthAdapter(http, store)
        adapter.signInWithGoogle()

        val seen = mutableListOf<String?>()
        adapter.subscribeToUser { seen.add(it?.displayName) }
        adapter.signOut()

        assertNull(store.current())
        assertEquals(listOf("Vgaka", null), seen)
    }

    @Test
    fun `subscribeToUser fires the current snapshot immediately`() {
        val adapter = SupabaseAuthAdapter(FakeHttp(), FakeStore())
        val seen = mutableListOf<String?>()
        adapter.subscribeToUser { seen.add(it?.displayName) }
        assertEquals(listOf<String?>(null), seen)
    }

    @Test
    fun `accessToken returns the stored JWT`() = runTest {
        val store = FakeStore(initial = session)
        val adapter = SupabaseAuthAdapter(FakeHttp(), store)
        assertEquals("jwt-abc", adapter.accessToken())
    }

    @Test
    fun `providerAccessToken proxies via http when JWT is available`() = runTest {
        val http = FakeHttp(refreshedToken = "ya29.fresh")
        val adapter = SupabaseAuthAdapter(http, FakeStore(initial = session))
        val token = adapter.providerAccessToken(ExternalProviderId.GoogleDrive)
        assertEquals("ya29.fresh", token)
    }

    @Test
    fun `providerAccessToken returns null without a session`() = runTest {
        val http = FakeHttp(refreshedToken = "ya29.fresh")
        val adapter = SupabaseAuthAdapter(http, FakeStore())
        assertNull(adapter.providerAccessToken(ExternalProviderId.GoogleDrive))
    }

    @Test
    fun `OAuthOutcome rejects blank reasons`() {
        assertThrows(IllegalArgumentException::class.java) { OAuthOutcome.Cancelled("  ") }
        assertThrows(IllegalArgumentException::class.java) { OAuthOutcome.Failed(" ", null) }
    }

    @Test
    fun `AuthResult rejects blank messages`() {
        assertThrows(IllegalArgumentException::class.java) { AuthResult.Cancelled("") }
        assertThrows(IllegalArgumentException::class.java) { AuthResult.Error("  ", null) }
    }

    @Test
    fun `StoredSession enforces invariants`() {
        assertThrows(IllegalArgumentException::class.java) {
            StoredSession(userId, "", "rt", 1)
        }
        assertThrows(IllegalArgumentException::class.java) {
            StoredSession(userId, "jwt", "rt", 0)
        }
    }
}

private class FakeHttp(
    val oauth: OAuthOutcome = OAuthOutcome.Cancelled("not started"),
    val profile: UserProfileRow? = null,
    val refreshedToken: String? = null,
) : SupabaseAuthHttp {
    val scopesRequested = mutableListOf<List<String>>()
    var signOutCalls = 0

    override suspend fun startGoogleOAuth(scopes: List<String>): OAuthOutcome {
        scopesRequested.add(scopes)
        return oauth
    }

    override suspend fun signOut() {
        signOutCalls += 1
    }

    override suspend fun fetchUserProfile(userId: UserId, accessToken: String): UserProfileRow? =
        profile

    override suspend fun refreshProviderToken(jwt: String, provider: ExternalProviderId): String? =
        refreshedToken
}

private class FakeStore(initial: StoredSession? = null) : SessionStore {
    private var stored: StoredSession? = initial
    override suspend fun current(): StoredSession? = stored
    override suspend fun save(session: StoredSession) { stored = session }
    override suspend fun clear() { stored = null }
}
